### PR TITLE
JS2 week2 homework added

### PR DESCRIPTION
**Pull request by vitalreddym**

### CODE REVIEW

File: javascript/javascript2/week2/index.js [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+const shortMovies = movies.filter(short => short.title.length <= 3);
+console.log(shortMovies);
+
+
+
+// 2. Movies with long title
+
+const longMovies = movies.filter(long => long.title.length >=20 );
+console.log(longMovies);
+
```
Comment (ncryer) [inline javascript/javascript2/week2/index.js:14]: Excellent!

File: javascript/javascript2/week2/index.js [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+
+// 3. Count the number of movies between 1980-1989 including both 
+
+const count = movies.filter(title => {
+    return title.year >= 1980 && title.year <= 1989;
+}).length;
```
Comment (ncryer) [inline javascript/javascript2/week2/index.js:21]: Excellent solution. A side note: the boolean condition can be done in one *expression*, so you could also write `movies.filter(title => title.year >= 1980 && title.year <= 1989)`, but that's a minor detail

File: javascript/javascript2/week2/index.js [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+// 6. Count the total number of movies based on keywords: Surfer, Alien or Benjamin
+
+const countKeyWords = movies.reduce((counter, movie) => {
+    if (movie.title.toLowerCase().includes ("surfer") ||
+        movie.title.toLowerCase().includes ("alien") ||
+        movie.title.toLowerCase().includes ("benjamin") )
+
+    counter += 1;
+    
+    return counter;
```
Comment (ncryer) [inline javascript/javascript2/week2/index.js:61]: Very good!!!

File: javascript/javascript2/week2/index.js [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+const titlesWithDuplicateWords = [];
+
+movies.map(movie => movie.title.split(" "))
+	.forEach(titleWord =>
+		titleWord.filter((word, index) =>
+			titleWord.indexOf(word) != index
+				? titlesWithDuplicateWords.push(titleWord)
+				: false
+		)
+	);
```
Comment (ncryer) [inline javascript/javascript2/week2/index.js:80]: Very inventive solution! Beautiful use of method chaining here, the gods of functional programming are surely smiling from above! 
 
However something is off about the logic, and I'm not entirely sure what it is. For one thing, we're not accounting for differences purely due to capitalization,so if you add `movie.title.toLowerCase().split(" ")` instead you end up with 236 titles. A number of those are also counted twice! 
 
Something to keep in mind is the intended usage of `map`, `filter` and `forEach`. `map` and `filter` are functions which take arrays as input and also output arrays, so it's good to think of them as singular "transformations" of an input array - and here we're using both methods as essentially for-loops, without really using the output. 
 
Here's my solution, which finds 204 movies: 
```js
const duplicatedTitles = movies.filter(movie => {
    const wordCount = {};
    movie.title.toLowerCase()
        .split(" ")
        .forEach(w => wordCount[w] = !wordCount[w] ? 1 : wordCount[w] + 1);
    return Object.keys(wordCount).some(key => wordCount[key] > 1);
});
```

Here I'm finding movies by filtering on the following way: for each movie title, keep a word count object that maps each word to an integer count. As we loop over the words in the (lowercased) title with `forEach`, we update the word count object by either initializing the word to 1, or incrementing it if it's already there. Then I return true if any of the words have a value greater than 1.

File: javascript/javascript2/week2/index.js [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+        if (movie.tag in count) {
+            count[movie.tag]++;
+        } 
+         else  {
+            count[movie.tag] = 1;
+        }
+        return count;
+    },
+	{ Good: 0, Average: 0, Bad: 0 }
+);
```
Comment (ncryer) [inline javascript/javascript2/week2/index.js:123]: Excellent use of reduce, well done

File: javascript/javascript2/week2/main.js [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+//Sort the products on rating
+const inputRating = document.querySelector(".input-rating");
+inputRating.addEventListener("input", function () {
+  const filterByRating = products.filter((enter) => {
+    return enter.rating >= inputRating.value;
+  });
+
+  listOfProducts.innerHTML = " ";
+  renderProducts(filterByRating);
+});
```
Comment (ncryer) [inline javascript/javascript2/week2/main.js:68]: Excellent, great work!

**Review by ncryer — APPROVED** [review]

File: javascript/javascript2/week2/index.js (hunk 2 above)
Comment (vitalreddym) [reply javascript/javascript2/week2/index.js:21]: Thanks for the explanation.

**Pull request merged**